### PR TITLE
scaler: Sync with upstream behavior

### PR DESCRIPTION
**Pull request by kargakis**
Also added some unit testing

Once @deads2k's Kube rebase is in, this will match upstream changes in the scaler.
@ironcladlou PTAL

### CODE REVIEW

File: pkg/deploy/scaler/scale_test.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+			count:     uint(3),
+			oc:        testclient.NewSimpleFake(deploytest.OkDeploymentConfig(1)),
+			kc:        ktestclient.NewSimpleFake(deploytest.MkDeploymentList(1)),
+			expected: []string{
+				"get-deploymentconfig",
+			},
+			kexpected: []string{
+				"get-replicationController",
+				"update-replicationController",
+			},
```
Comment (ironcladlou) [inline pkg/deploy/scaler/scale_test.go]: I find this kind of expectation setting extremely hard to understand and maintain, but I'm not clear on the current preferences upstream. In the past I thought there was buy-in on using injected function mocks over the fake clients for code like this, but at least @derekwaynecarr seemed opposed. I don't mind letting it in here, just stating my opinion that I'd prefer a different style.

Comment (liggitt) [inline pkg/deploy/scaler/scale_test.go]: I like the expected operations, but this is the first I've seen it in action with two clients... that's pretty hard to follow

Comment (kargakis) [inline pkg/deploy/scaler/scale_test.go]: > since there's no overall ordering

Yea, that suxx. We should unify our mock client with the Kubernetes one

Comment (ironcladlou) [inline pkg/deploy/scaler/scale_test.go]: My concerns are more about: You've got some interface method you're testing (`Scale`) and the behavior you're defining for the mock dependencies of the scaler are expressed as an uncommented, uncorrelated list of generic API calls (`get`, `update`, etc.)

What line of code the test scaler does the 4th entry in that list (e.g., a `get` amongst a chain of 3 `get` calls) actually correspond to? You have to do like a linear scan of the code every time you want to figure out what the mock operation actually corresponds to. Then apply that to function calls, other dependencies throughout the code, etc. This gets crazy very quickly as the code grows. When you add a new piece of logic to the scaler that interfaces with a mocked dependency, now you have to go back to your linear scanning to figure out how to inject a new out-of-context mock method into that list, compounding the issue every time.

Compare this with a mocking method whereby `Scaler.updateReplicationController` is mocked directly with a function injection. You know immediately how the mock is set up and where it's called from in code. The order of this call in relation to other mocked calls is almost always irrelevant (and when it is relevant, the ordering assertions can be expressed with code that relates other clearly mock injection points in the test code).

Is the order of the mock calls really as significant as these lists imply? I don't yet see how that property justifies the cognitiive overhead of using them for the majority of cases I've seen.

**Comment by kargakis** [comment]
@ironcladlou can we move on for now with this? My changes to the testclient aren't going in anytime soon...

File: pkg/deploy/util/util.go [hunk 2]
```diff
@@ -163,6 +164,23 @@ func MakeDeployment(config *deployapi.DeploymentConfig, codec runtime.Codec) (*a
...
+// mkdeployment returns a deployment
+// Used for testing
+func mkdeployment(version int) api.ReplicationController {
+	deployment, _ := MakeDeployment(deploytest.OkDeploymentConfig(version), api.Codec)
+	return *deployment
+}
+
+// MkDeploymentList returns a list of deployments
+// Used for testing
+func MkDeploymentList(versions ...int) *api.ReplicationControllerList {
```
Comment (ironcladlou) [inline pkg/deploy/util/util.go]: This belongs in `pkg/deploy/api/test/ok.go`, not the public utility.

**Comment by ironcladlou** [comment]
>@ironcladlou can we move on for now with this? My changes to the testclient aren't going in anytime soon...

For sure, I didn't mean to block on the test discussion. Sorry!

File: pkg/deploy/util/util.go (hunk 2 above)
Comment (kargakis) [inline pkg/deploy/util/util.go]: I had moved those funcs there originally but I was getting cyclic imports. I will move them back just to refresh my memory. Most probably will move them back in the tests

Comment (kargakis) [inline pkg/deploy/util/util.go]: Hmm, they work fine, not sure what I did a couple of days ago... PTAL

**Comment by ironcladlou** [comment]
LGTM [merge]

**Comment by kargakis** [comment]
```
# github.com/openshift/origin/pkg/deploy/util
import cycle not allowed in test
package github.com/openshift/origin/pkg/deploy/util (test)
	imports github.com/openshift/origin/pkg/deploy/api/test
	imports github.com/openshift/origin/pkg/deploy/util

FAIL	github.com/openshift/origin/pkg/deploy/util [setup failed]
```
util/util_test.go imports deploy/api/test and test/ok.go imports deploy/util. Should we move `MakeDeployment` from util to another place or leave the test functions I have refactored as are and duplicate them in the scaler tests?

**Comment by ironcladlou** [comment]
Just move the test functions out of `ok.go` and into the scaler test. This kind of tiny test helper duplication doesn't really bother me.
